src/inmanta/execute/runtime.py

```diff
@@ -597,14 +598,12 @@ def _notify_listeners(self, value: ListValue, location: Location) -> None:
         results.
         """
         assert self._listeners is not None
-        for listener in list(self._listeners.keys()):
+        for listener in self._listeners:
             done: bool = listener.receive_result(value, location)
             if done:
-                if listener.pure_gradual():
-                    self._nb_pure_listeners -= 1
-                # keep memory footprint minimal
-                del self._listeners[listener]
-                self._done_listeners += 1
+                # TODO: comments
+                if not listener.pure_gradual():
+                    self._nb_pure_listeners += 1
```

Comment:
TODO: create a test case that catches this bug: listener keeps getting results => count keeps increasing.